compiler/tests-common/tests/org/jetbrains/kotlin/codegen/AbstractCheckLocalVariablesTableTest.kt

```diff
@@ -120,13 +134,296 @@ abstract class AbstractCheckLocalVariablesTableTest : CodegenTestCase() {
+                        else -> throw Exception("Unsupported var instruction: $opcode")
+                    }
+                }
+            }
+
+            // Visitor that builds a control-flow graph of a method. The control-flow graph
+            // consists of [BasicBlock]s containing the local load and store instructions
+            // and the locals table at each basic block. When the visitor ends, the type of
+            // local slots are computed based on instructions in the code and are checked
+            // for consistency with information in the locals table.
+            class Visitor : ClassVisitor(Opcodes.API_VERSION) {
+                var methodFound = false
+                var skipValidation = false
+
+                var entryBlock = BasicBlock()
+                var allBlocks: MutableSet<BasicBlock> = mutableSetOf(entryBlock)
+                var currentBlock = entryBlock
+                var labelToBlock: MutableMap<Label, BasicBlock> = mutableMapOf()
+                val currentLocalsTable: MutableMap<Int, String> = mutableMapOf()
+
+                private fun ensureBlock(label: Label): BasicBlock {
+                    return if (labelToBlock.containsKey(label)) {
+                        labelToBlock[label]!!
+                    } else {
+                        val result = BasicBlock()
+                        allBlocks.add(result)
+                        labelToBlock[label] = result
+                        result
+                    }
+                }
+
+                private fun recordLocalTypesForParameters(access: Int, desc: String) {
+                    val localsAtEntry = entryBlock.localsAtEntry
+                    var parameterIndex = 0
+                    if (access and Opcodes.ACC_STATIC == 0) {
+                        localsAtEntry[parameterIndex++] = "Ljava/lang/Object;"
+                    }
+                    Type.getMethodType(desc).argumentTypes.forEach {
+                        when (it) {
+                            // Bytecode instructions use iload/istore for all of these so we do not distinguish.
+                            Type.BOOLEAN_TYPE,
+                            Type.CHAR_TYPE,
+                            Type.BYTE_TYPE,
+                            Type.SHORT_TYPE,
+                            Type.INT_TYPE -> localsAtEntry[parameterIndex++] = "I"
+
+                            Type.FLOAT_TYPE-> localsAtEntry[parameterIndex++] = "F"
+
+                            Type.DOUBLE_TYPE -> {
+                                localsAtEntry[parameterIndex++] = "D"
+                                // A double takes up two slots.
+                                parameterIndex++
+                            }
+                            Type.LONG_TYPE -> {
+                                localsAtEntry[parameterIndex++] = "J"
+                                // A double takes up two slots.
+                                parameterIndex++
```

Comment:
All increments could be changed to 'parameterIndex += it.size'

Reply:
and then localsAtEntry[...] could be pushed up from when

Reply:
Done, thanks!